mathinspector/__main__.py

```diff
@@ -1,6 +1,67 @@
-from .app import main
+"""
+Math Inspector: a visual programming environment for scientific computing
+Copyright (C) 2021 Matt Calhoun
 
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
 
-if __name__ == '__main__':
-	# mp.freeze_support() #required for windows
-	main()
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""
+import sys
+from . import main
+
+if __name__ == "__main__":
+	args = []
+	kwargs = {}
+	i = 1
+
+	while i < len(sys.argv):
+		arg = sys.argv[i]
+		if arg[:2] != "--":
+			args.append(sys.argv[i])
+		elif "=" in arg:
+			key, value = arg[2:].split("=")
+			kwargs[key] = value
+		elif i + 1 < len(sys.argv) and sys.argv[i+1] and sys.argv[i+1][:2] != "--":
+			kwargs[arg[2:]] = sys.argv[i+1]
+			i += 1
+		else:
+			kwargs[arg[2:]] = True
+		i += 1
+
+	if "help" in kwargs and kwargs["help"] is True:
+		print ("""usage: mathinspector [files] ... [options] ... 
+
+Options and arguments:
+[files]...       : a list of files with either a .math or .py extension.  The
+                  .math file will be loaded, and all of the .py files will be
+                  added to the current project
+
+--help obj       : view the documentation for obj as if you called help(obj)
+                  in mathinspector
+
+--new            : starts a new project and resets the state of the app.
+                  This flag will overwrites the autosave file with a blank
+                  file
+
+--disable[=opts] : mathinspector overrides a lot of builtins (e.g. help(),
+                  print()) which may cause undesirable behaviour, this flag
+                  disables these features
+                    opts are 'print', 'traceback', 'stderr'
+                    e.g. mathinspector --disable=print,stderr
+
+--debug          : prints log messages to the command line used to launch the
+                  app.  Useful for debugging issues when something isn't
+                  working properly, or while working on the mathinspector
+                  source code
+		""")
```

Comment:
this logic should be in the `main()` function itself

Reply:
I disagree, as I previously explained, I had been planning to use the `__main__` file exactly like this.  Why do you think it should be in `main()`?

Reply:
I thought this was an interesting comment about putting the sys.argv logic in main()

If a user has a script `test.py` that has `mathinspector.main()`, and they call this script with argv's
```
$ python test.py --somearg
```
then this will cause the value of `sys.argv` in the `main()` function to get bad data and the app will fail to launch.  Which would be a bug imo.

Reply:
there are lots of reasons but it's easiest to ask the other way around, really - why didn't you? I presume you looked at a few established python libraries and so should've seen they put the `sys.argv` parsing login in a function (of the world's top 15 out of ~330k ranked Python packages, 3 implement `__main__.py`, and they all use a function: https://github.com/pytest-dev/pytest/blob/master/src/pytest/__main__.py, https://github.com/tqdm/tqdm/blob/master/tqdm/__main__.py, https://github.com/ipython/ipython/blob/master/IPython/__main__.py).

Reply:
er, no, a user calling `mathinspector.main()` intends to forward `sys.argv` to mathinspector. If they intend otherwise they need to use `mathinspector.main(argv=[])` instead. Once again, check other packages for convention :)

Reply:
> it's easiest to ask the other way around, really - why didn't you?

There are lots of reasons.  First of all, I wanted people to be able to use
```python
import mathinspector
mathinspector.main()
```
in their own source code files.  I wanted it to be the same `main()` function the app uses for the cli but to accept python style `*args, **kwargs`.  I wanted to keep the logic for parsing argv in __main__ so that in the future that file can grow much larger and be completely dedicated to a robust cli for mathinspector, and to do so in a way that every new addition to the cli is automatically applied to not only the import interface, but also both the in-app GUI and the math inspector source code files themselves.

It just didn't seem like there was anything wrong with using the `__main__.py` file in this way.  I am very happy with it.

I understand that other python packages don't use `__main__` like this, but I have a lot more reasons why I wanted to keep it like this and it's the result of careful consideration.

What I would like to know is: why do 12 out of the top 15 python packages feel it's important to not use `__main__` in the way I did?

Reply:
you're talking about a `main-run` pattern - I can push this in a different branch.

Reply:
tbh, that bit with how `__main__` parses sys.argv and passes it as `*args, **kwargs` to the `main()` function in `__init__` is something I was super proud of and until you said you wanted to move it, I had been eagerly awaiting your comment on that because I thought it was that good.

I guess I kind of feel strongly that I want to keep it, and also since I like that pattern so much and just made it up, I'd like to understand more about what the top packages consider to be best practices.

Reply:
In this case it's not what you wrote, it's just where you wrote it that's bad. See #43.

Reply:
I believe an important part of working with other people is I don't always get to have my way.  In this case, it's truly very important to me that you feel good about whatever resolution we come up with.  I also very much want to learn from you more about the higher levels of python.

I have been programming for my entire life, and have come to the strong conclusion that having a dedicated file in the root of the most important folder in a project that does nothing but import a function from another file and call it once is a bad practice.

I disagree that where I put it was bad.  I claim that the so-called "best practice" of having a blank `__main__` file is in fact misusing the original design of python and why there is a `__main__` file.  It's a way of trying to work against python imo.

Reply:
> If they intend otherwise they need to use mathinspector.main(argv=[]) instead. 

I strongly disagree with the concept that `mathinspector.main(argv=[])` is a preferable interface to `mathinspector.main(*args, **kwargs)` for calling a function named `main` after doing `import mathinspector` in either a file or from the standard python interpreter.

> Once again, check other packages for convention :)

You keep on making references to authority, and with all due respect to your very high skill level at programming; I assert that I am an expert on the philosophy of writing good code with literally a lifetime of experience across over 10 programming languages, over 30 years, and I have been the CTO of multiple successful companies whose technological infrastructure I created from scratch and personally maintained while working with a team of people who reported to me.

I have asked you multiple times to justify your position because I want to learn from you, because I value and respect your opinion.  So "once again", please justify your statement that the "where I put it was bad".

Reply:
Ah sorry but I wasn't offering an *explanation* but a *reason* ("everyone does it so you should too" - if viewed as an *explanation* this would of course classify as a logical fallacy).

I could spend about half an hour writing a comprehensive explanation which would be the beginnings of a subsection in a book on Python which I admittedly should write at some point in the future. However for now, to be blunt if I'm not being paid to do something then it's a hobby. And if it's a hobby I don't owe anyone an explanation. However I did anticipate a reasonable view that *reason sans explanation* does not warrant inclusion in `master` and thus - rather than commit here - opened a separate PR which you are free to ignore.

I might find time to respond to some of your review comments in #43 but I'm afraid unlikely to delve into comparisons of design patterns and APIs.

Reply:
I would like to merge this PR into master as is, and close #43 in a few days unless you reply to my review comments.  Does that sound good to you?

It's not necessary to reply to any of those review comments.  If you do, I would be very interested in having a philosophical discussion about the specific technical details of what's going on here.

> to be blunt if I'm not being paid to do something then it's a hobby. And if it's a hobby I don't owe anyone an explanation

Well said.  I am also not being paid for this, although it's more than just a hobby for me, since I am so personally invested in this project at this point.  If anything, I feel like I am in your debt for your amazing contribution to the python package structure.  It seems fair to heavily weight your opinion when making decisions moving forward for this reason.

If you claim something is True, that's generally good enough for me, and I don't need an explanation, nor think you owe me one.  In this case, we are talking about something I had been planning for 6 months and have carefully considered and researched.  I hope you have enjoyed working on this project, and that whatever reasons you have in the first place for being the first contributor, that we can build on those with this project so it's not just mine anymore, but belongs to the entire community.

> I could spend about half an hour writing a comprehensive explanation which would be the beginnings of a subsection in a book on Python which I admittedly should write at some point in the future

I was asking for an explanation because I would like to read that book!

> I did anticipate a reasonable view that reason sans explanation does not warrant inclusion in master

Thanks for this.